Failed tests (10599): TestGossip{Peerings,Restart}: newRaft: applied(10) is out of range (...)

The following test appears to have failed:

[#10599](https://circleci.com/gh/cockroachdb/cockroach/10599):

```
I0105 14:29:46.233607 1 server/context.go:181  initialized 1 storage engine(s)
I0105 14:29:46.233663 1 cli/start.go:176  starting cockroach cluster
I0105 14:29:46.236317 1 storage/engine/rocksdb.go:106  opening rocksdb instance at "/data2.0"
I0105 14:29:46.237581 1 gossip/gossip.go:738  starting client to roach0.local:26257
E0105 14:29:46.253311 1 raft/log.go:205  [group 3] applied(10) is out of range [prevApplied(27), committed(30)]
panic: [group 3] applied(10) is out of range [prevApplied(27), committed(30)]

goroutine 1 [running]:
github.com/cockroachdb/cockroach/storage.(*raftLogger).Panicf(0xc8203e0b00, 0x160a000, 0x3c, 0xc8203cfaa0, 0x3, 0x3)
	/go/src/github.com/cockroachdb/cockroach/storage/raft.go:124 +0x1a2
github.com/coreos/etcd/raft.(*raftLog).appliedTo(0xc8203e80e0, 0xa)
	/go/src/github.com/coreos/etcd/raft/log.go:205 +0x1f7
github.com/coreos/etcd/raft.newRaft(0xc820256828, 0x7f8cd956e7e8)
	/go/src/github.com/coreos/etcd/raft/raft.go:224 +0x939
github.com/coreos/etcd/raft.NewRawNode(0xc820256828, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/coreos/etcd/raft/rawnode.go:76 +0xba
I0105 14:29:46.224917 18465 acceptance/gossip_peerings_test.go:180  waiting for gossip to be connected
E0105 14:29:46.225906 18465 acceptance/cluster/localcluster.go:406  node=2 status=restart
E0105 14:29:46.313702 18465 acceptance/cluster/localcluster.go:406  node=2 status=die
I0105 14:30:06.351488 18465 acceptance/cluster/localcluster.go:533  stopping
I0105 14:30:06.910577 18465 acceptance/cluster/localcluster.go:572  node 2: stderr at /tmp/circle-artifacts.Tft7Deb/acceptance/roach2.local/stderr.2016-01-05T14_30_06Z.log
--- FAIL: TestGossipRestart (35.99s)
	testing.go:145: acceptance/gossip_peerings_test.go:68: condition failed to evaluate within 20s: acceptance/gossip_peerings_test.go:63: node 0: acceptance/gossip_peerings_test.go:82: expected 3 peers, found 2
	localcluster.go:510: unexpected extra event &{2 die} (after [{0 die} {1 die} {2 die} {0 restart} {1 restart} {2 restart}])
=== RUN   TestPut
I0105 14:30:07.878445 18465 acceptance/cluster/localcluster.go:213  started docker-spy: 172.17.0.33
I0105 14:30:07.878470 18465 acceptance/cluster/localcluster.go:220  initializing cluster
I0105 14:30:09.021700 18465 acceptance/cluster/localcluster.go:331  creating ca (1024bit) in: /home/ubuntu/cockroach/.localcluster.certs.919478579
I0105 14:30:09.062160 18465 acceptance/cluster/localcluster.go:336  creating node (1024bit) certs in: /home/ubuntu/cockroach/.localcluster.certs.919478579
I0105 14:30:10.468895 18465 acceptance/cluster/localcluster.go:389  *** started roach0.local ***
  ui:    https://0.0.0.0:32790
  trace: https://0.0.0.0:32790/debug/requests
--
--- SKIP: TestBuildBabyCluster (0.00s)
	terraform_test.go:32: only enabled during testing
=== RUN   TestFiveNodesAndWriters
--- SKIP: TestFiveNodesAndWriters (0.00s)
	main_test.go:56: running in docker mode
FAIL
ok github.com/cockroachdb/cockroach/acceptance 1337s

```
Please assign, take a look and update the issue accordingly.

--- a/acceptance/chaos_test.go
+++ b/acceptance/chaos_test.go
@@ -39,7 +39,6 @@ import (
 // independent keys, while nodes are being killed and restarted continuously.
 // The test measures not write performance, but cluster recovery.
 func TestChaos(t *testing.T) {
-	t.Skip("#3640")
 	c := StartCluster(t)
 	defer c.AssertAndStop(t)
 

--- a/acceptance/gossip_peerings_test.go
+++ b/acceptance/gossip_peerings_test.go
@@ -102,7 +102,6 @@ func hasClusterID(infos map[string]interface{}) error {
 }
 
 func TestGossipPeerings(t *testing.T) {
-	t.Skip("#3611")
 	c := StartCluster(t)
 	defer c.AssertAndStop(t)
 	num := c.NumNodes()
@@ -141,7 +140,6 @@ func TestGossipPeerings(t *testing.T) {
 // re-bootstrapped after a time when all nodes were down
 // simultaneously.
 func TestGossipRestart(t *testing.T) {
-	t.Skip("#3611")
 	// This already replicates the first range (in the local setup).
 	// The replication of the first range is important: as long as the
 	// first range only exists on one node, that node can trivially

--- a/storage/replica.go
+++ b/storage/replica.go
@@ -1069,21 +1069,26 @@ func (r *Replica) handleRaftReady() error {
 	r.Unlock()
 	logRaftReady(r.store.StoreID(), r.RangeID, rd)
 
+	batch := r.store.Engine().NewBatch()
+	defer batch.Close()
 	if !raft.IsEmptySnap(rd.Snapshot) {
-		if err := r.applySnapshot(rd.Snapshot); err != nil {
+		if err := r.applySnapshot(batch, rd.Snapshot); err != nil {
 			return err
 		}
 		// TODO(bdarnell): update coalesced heartbeat mapping with snapshot info.
 	}
 	if len(rd.Entries) > 0 {
-		if err := r.append(rd.Entries); err != nil {
+		if err := r.append(batch, rd.Entries); err != nil {
 			return err
 		}
 	}
 	if !raft.IsEmptyHardState(rd.HardState) {
-		if err := r.setHardState(rd.HardState); err != nil {
+		if err := r.setHardState(batch, rd.HardState); err != nil {
 			return err
 		}
+	}
+	if err := batch.Commit(); err != nil {
+		return err
 	}
 
 	for _, msg := range rd.Messages {

--- a/storage/replica_raftstorage.go
+++ b/storage/replica_raftstorage.go
@@ -395,13 +395,10 @@ func (r *Replica) Snapshot() (raftpb.Snapshot, error) {
 }
 
 // append the given entries to the raft log.
-func (r *Replica) append(entries []raftpb.Entry) error {
+func (r *Replica) append(batch engine.Engine, entries []raftpb.Entry) error {
 	if len(entries) == 0 {
 		return nil
 	}
-	batch := r.store.Engine().NewBatch()
-	defer batch.Close()
-
 	for _, ent := range entries {
 		err := engine.MVCCPutProto(batch, nil, keys.RaftLogKey(r.RangeID, ent.Index),
 			roachpb.ZeroTimestamp, nil, &ent)
@@ -424,11 +421,10 @@ func (r *Replica) append(entries []raftpb.Entry) error {
 	if err := setLastIndex(batch, r.RangeID, lastIndex); err != nil {
 		return err
 	}
-	if err := batch.Commit(); err != nil {
-		return err
-	}
+	batch.Defer(func() {
+		atomic.StoreUint64(&r.lastIndex, lastIndex)
+	})
 
-	atomic.StoreUint64(&r.lastIndex, lastIndex)
 	return nil
 }
 
@@ -464,7 +460,7 @@ func (r *Replica) updateRangeInfo() error {
 }
 
 // applySnapshot updates the replica based on the given snapshot.
-func (r *Replica) applySnapshot(snap raftpb.Snapshot) error {
+func (r *Replica) applySnapshot(batch engine.Engine, snap raftpb.Snapshot) error {
 	snapData := roachpb.RaftSnapshotData{}
 	err := proto.Unmarshal(snap.Data, &snapData)
 	if err != nil {
@@ -473,10 +469,15 @@ func (r *Replica) applySnapshot(snap raftpb.Snapshot) error {
 
 	rangeID := r.RangeID
 
-	// First, save the HardState.  The HardState must not be changed
-	// because it may record a previous vote cast by this node.
+	// First, save the HardState. The HardState must not be changed
+	// because it may record a previous vote cast by this node. This is
+	// usually unnecessary because a snapshot is nearly always
+	// accompanied by a new HardState which incorporates both our former
+	// state and new information from the leader, but in the event that
+	// the HardState has not changed, we want to use our own previous
+	// HardState and not one that was transmitted via the snapshot.
 	hardStateKey := keys.RaftHardStateKey(rangeID)
-	hardState, _, err := engine.MVCCGet(r.store.Engine(), hardStateKey, roachpb.ZeroTimestamp, true /* consistent */, nil)
+	hardState, _, err := engine.MVCCGet(batch, hardStateKey, roachpb.ZeroTimestamp, true /* consistent */, nil)
 	if err != nil {
 		return err
 	}
@@ -484,11 +485,8 @@ func (r *Replica) applySnapshot(snap raftpb.Snapshot) error {
 	// Extract the updated range descriptor.
 	desc := snapData.RangeDescriptor
 
-	batch := r.store.Engine().NewBatch()
-	defer batch.Close()
-
 	// Delete everything in the range and recreate it from the snapshot.
-	iter := newReplicaDataIterator(&desc, r.store.Engine())
+	iter := newReplicaDataIterator(&desc, batch)
 	defer iter.Close()
 	for ; iter.Valid(); iter.Next() {
 		if err := batch.Clear(iter.Key()); err != nil {
@@ -545,38 +543,36 @@ func (r *Replica) applySnapshot(snap raftpb.Snapshot) error {
 		return err
 	}
 
-	if err := batch.Commit(); err != nil {
-		return err
-	}
+	batch.Defer(func() {
+		// Update the range stats.
+		r.stats.Replace(newStats)
 
-	// Update the range stats.
-	r.stats.Replace(newStats)
+		// As outlined above, last and applied index are the same after applying
+		// the snapshot.
+		atomic.StoreUint64(&r.lastIndex, snap.Metadata.Index)
+		atomic.StoreUint64(&r.appliedIndex, snap.Metadata.Index)
 
-	// As outlined above, last and applied index are the same after applying
-	// the snapshot.
-	atomic.StoreUint64(&r.lastIndex, snap.Metadata.Index)
-	atomic.StoreUint64(&r.appliedIndex, snap.Metadata.Index)
+		// Atomically update the descriptor and lease.
+		if err := r.setDesc(&desc); err != nil {
+			panic(err)
+		}
+		// Update other fields which are uninitialized or need updating.
+		// This may not happen if the system config has not yet been loaded.
+		// While config update will correctly set the fields, there is no order
+		// guarangee in ApplySnapshot.
+		// TODO: should go through the standard store lock when adding a replica.
+		if err := r.updateRangeInfo(); err != nil {
+			panic(err)
+		}
 
-	// Atomically update the descriptor and lease.
-	if err := r.setDesc(&desc); err != nil {
-		return err
-	}
-	// Update other fields which are uninitialized or need updating.
-	// This may not happen if the system config has not yet been loaded.
-	// While config update will correctly set the fields, there is no order
-	// guarangee in ApplySnapshot.
-	// TODO: should go through the standard store lock when adding a replica.
-	if err := r.updateRangeInfo(); err != nil {
-		return err
-	}
-
-	atomic.StorePointer(&r.lease, unsafe.Pointer(lease))
+		atomic.StorePointer(&r.lease, unsafe.Pointer(lease))
+	})
 	return nil
 }
 
 // setHardState persists the raft HardState.
-func (r *Replica) setHardState(st raftpb.HardState) error {
-	return engine.MVCCPutProto(r.store.Engine(), nil, keys.RaftHardStateKey(r.RangeID),
+func (r *Replica) setHardState(batch engine.Engine, st raftpb.HardState) error {
+	return engine.MVCCPutProto(batch, nil, keys.RaftHardStateKey(r.RangeID),
 		roachpb.ZeroTimestamp, nil, &st)
 }
 

